adr/ADR-15.md

```diff
@@ -0,0 +1,233 @@
+
+## Design
+
+### Creation
+
+The library should have API(s) that allow(s) creation of a JetStream subscription. Depending on the language, there may be various type of subscriptions (like Channel based in Go), or with various options such as Queue subscriptions, etc... There can be different APIs or one with a set of options.
+
+#### Configuration
+
+When creating a subscription, the user can provide several things:
+
+- A subject that is in some cases required (see "Stream name" section below)
+- An optional stream name
+- An optional consumer name
+- An optional JS consumer configuration (see below), if the library is expected to create the JS consumer
+- A Queue name if this subscription is meant to be a queue subscription
+- Indication if this is a "Pull" subscription or not
+- An optional way of indicating the the given subject will be used to create the internal subscription, by-passing any lookup/consumer creation. For the purpose of this document, let's call this option `SubjectIsDelivery`.
+
+Some misconfiguration should be checked by the subscribe API and return an error outright.
+
+For instance:
+- If a queue name is provided, configuring heartbeat or flow control is a mistake, since the server would send those message to random members.
+- If no subject is provided, along with no stream name or the `SubjectIsDelivery` option set, the library won't be able to locate/create the consumer.
+- For pull subscriptions:
+	- ack policy of "none" or "all" is an error.
+	- setting a deliver subject or using the `SubjectIsDelivery` option is an error.
+
+At the time of this writing, the consumer configuration object looks like this (ordred alphabetically, not as seen in js.go):
+
+```go
+type ConsumerConfig struct {
+	AckPolicy       AckPolicy     `json:"ack_policy"`
+	AckWait         time.Duration `json:"ack_wait,omitempty"`
+	DeliverGroup    string        `json:"deliver_group,omitempty"`
+	DeliverPolicy   DeliverPolicy `json:"deliver_policy"`
+	DeliverSubject  string        `json:"deliver_subject,omitempty"`
+	Description     string        `json:"description,omitempty"`
+	Durable         string        `json:"durable_name,omitempty"`
+	FilterSubject   string        `json:"filter_subject,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	MaxDeliver      int           `json:"max_deliver,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`
+	OptStartSeq     uint64        `json:"opt_start_seq,omitempty"`
+	OptStartTime    *time.Time    `json:"opt_start_time,omitempty"`
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	ReplayPolicy    ReplayPolicy  `json:"replay_policy"`
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+}
+```
+
+#### Stream name
```

Comment:
Can you start with a section describing the `SubjectIsDelivery` option? I have no context at this point to what it means.

Reply:
I have created a section before "Stream name" that describes this option and gives some details. Hope this clarifies it.